enhancements/installer/pre-flight-validations.md

```diff
@@ -0,0 +1,194 @@
+---
+title: pre-flight-validations
+authors:
+  - "@mandre"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-05-18
+last-updated: 2020-05-18
+status: implementable
+---
+
+# Pre-flight validations
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+- Will the validations be run automatically on `cluster create` or will it be
+  an explicit action?
+- If explicit action, how will the validations be enabled? E.g. adding
+  a `--dry-run` option to the installer vs. a separate subcommand or even
+  a separate binary.
+- When enabling the validations, will it also perform the installation or
+  simply run the validations?
+- Can the installer override validations that failed?
+- How are failures and warnings reported to the user?
+
+## Summary
+
+One of the guiding principles of OCP 4 is that the installation should always
+succeed. This goal is relatively easy to implement for public cloud platforms
+where we can make assumptions about services being available, or performance
+meeting requirements, however this is not the case with private clouds where
+each cloud is unique.
+
+It is currently not possible to tell with confidence if an installation will be
+successful or not for the following platforms:
+- BareMetal
+- OpenStack
+- oVirt
+- vSphere
+
+For this purpose, we propose to implement a framework allowing the installer to
+run pre-flight validations in order to certify that all pre-requisites are met
+to successfully install OpenShift in the selected environment.
+
+## Motivation
+
+This section is for explicitly listing the motivation, goals and non-goals of
+this proposal. Describe why the change is important and the benefits to users.
+
+### Goals
+
+By having an automated way to identify potential issues early, before the
+deployment even started, we want to reduce wasted time and resources, customer
+escalations, and improve the perception of OpenShift deployments on-premise.
+
+The installer already performs some validations:
+- checks all required fields are set and that the data is in the right format
+- some basic validation, such as networks do not overlap
+
+However, this doesn't check that the environment is suitable to install OpenShift:
+- pull secret is valid to fetch the container images
+- the tenant has adequate quota and the flavors' specifications are within the
+  recommended ranges.
+- for user-provided networks, check the subnets have a DHCP server and valid
+  DNS to reach the cloud's enpoints
```

Comment:
What's a "user-provided" network? Does this machine mean machine networks? How do you validate DHCP works? What are cloud endpoints? The metadata URI?

Reply:
nit: endpoint

Reply:
"user-provided" refers to networks not created by the installer.

https://github.com/openshift/installer/blob/master/docs/user/aws/customization.md#installing-to-existing-vpc--subnetworks
https://github.com/openshift/installer/blob/master/docs/user/gcp/customization.md#installing-to-existing-networks--subnetworks

In OpenStack platform, we also allow passing the machine network with `machinesSubnet` and setting additional networks to the VMs with `additionalNetworkIDs`, for multi-NIC VMs.

The cloud endpoints refers to the URLs you use to talk to your cloud's API. I'll expend a bit this section, hopefully clarifying things.